packages/webapp/src/components/Task/AddSoilAmendmentProducts/QuantityApplicationRate/useQuantityApplicationRate.ts

```diff
@@ -0,0 +1,172 @@
+          .from('kg/m2') // database unit weight / database unit area
+          .to('kg/ha'), // database unit application_rate_weight
+        { shouldValidate: true },
+      );
+    } else if ((volume || volume === 0) && !isWeight && volume_unit && application_area) {
+      setValue(
+        TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_VOLUME,
+        convert(volume / application_area)
+          .from('l/m2') // database unit volume / database unit area
+          .to('l/ha'), // database unit application_rate_volume
+        { shouldValidate: true },
+      );
+    }
+  };
+
+  const updateQuantity = () => {
+    const application_rate_weight = getValues(TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_WEIGHT);
+    const application_rate_volume = getValues(TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_VOLUME);
+    const application_area = getValues(TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED);
+    if (
+      isWeight &&
+      (application_rate_weight || application_rate_weight === 0) &&
+      application_rate_weight_unit &&
+      application_area
+    ) {
+      setValue(
+        TASK_PRODUCT_FIELD_NAMES.WEIGHT,
+        convert(application_rate_weight).from('kg/ha').to('kg/m2') * application_area,
+        { shouldValidate: true },
+      );
+    } else if (
+      !isWeight &&
+      (application_rate_volume || application_rate_volume === 0) &&
+      application_rate_volume_unit &&
+      application_area
+    ) {
+      setValue(
+        TASK_PRODUCT_FIELD_NAMES.VOLUME,
+        convert(application_rate_volume).from('l/ha').to('l/m2') * application_area,
+        { shouldValidate: true },
+      );
+    }
+  };
+
+  /* Trigger recalculation when units are changed */
+  useEffect(() => {
+    updateApplicationRate();
+  }, [weight_unit, volume_unit]);
+
+  useEffect(() => {
+    updateQuantity();
+  }, [application_rate_weight_unit, application_rate_volume_unit]);
+
+  /* For the preview string, replicate the conversion the unit component would do if the value had been displayed in a <Unit /> rather than a string. However, don't update upon changes to application_area_unit beyond the initial undefined > defined change */
+  const [previewStringValue, setPreviewStringValue] = useState<number | null>(null);
+  const [previewStringUnit, setPreviewStringUnit] = useState<string | null>(null);
+
+  useEffect(() => {
+    if (application_area_unit && previewStringValue === null) {
+      /* if the user-selected unit is in the wrong system (e.g. metric stored value but farm on imperial), use the <Unit /> converted unit */
+      const unit = location_area[system].units.includes(total_area_unit)
+        ? getUnitOptionMap()[total_area_unit]
+        : application_area_unit;
+
+      const value =
+        total_area &&
+        roundToTwoDecimal(
+          convertFn(location_area[system], total_area, location_area.databaseUnit, unit.value),
+        );
+
+      setPreviewStringValue(value);
+      setPreviewStringUnit(unit.label);
```

Comment:
Is this useEffect and state for previewStrings necessary? If not then I think it can be just calculated on render and if doing the calculation on every render is expensive then useMemo can be used?

Reply:
Ummm, it wasn't about expensive calculation, but rather to freeze the value of the preview string unit to the user-selected database value, while the application area unit should adapt dynamically based on the calculation. There might be a better way but I think it's kind of low on the priority list right now as something to optimize.

Reply:
I see what you mean, maybe I'm missing something but I just was curious why not just compute the previewStringValue and unit without using state? For example in the component render, just do 

```
const previewStringValue = total_area &&
   roundToTwoDecimal(
     convertFn(location_area[system], total_area, location_area.databaseUnit, unit.value),
   );
```

and pass that as a prop to` AreaApplicationSummary`

My reasoning for this is that with the useEffect, its triggering an unnecessary extra render. For instance, first render prevString is null, then immediately it renders again from the useEffect with setState, and also the same thing if the useEffect dependencies change.

Reply:
In your code example above, where are you getting the value of `unit` from? Still from watching the value of `application_area_unit`?

Reply:
Yes unit is exactly the same as its defined:

 ```
 const unit =
    total_area_unit && location_area[system].units.includes(total_area_unit)
      ? getUnitOptionMap()[total_area_unit]
      : application_area_unit;
```

Reply:
So I'm not sure I love the idea of the calculations being re-run on every render, but as I said above the string value + unit need to be frozen. If written like this, when the `application_area_unit` changes (e.g. you can test it in the Storybook story for imperial by setting the percentage to a low value), the string will update as well. Ideally I would like to memoize the calculation and _exclude_ `application_area_unit` from the dependencies, but, as you mentioned, it's initially `undefined`.

Maybe a ref instead of component state would have been a cleaner way to make sure the values are only set one time?

Reply:
I’m not sure about ref but I tried this out on my local and was able to achieve the desired result (let me know if it doesn't work). I put the calculations inside useMemo to avoid running on each render. Now as you mentioned it needs to be frozen and that’s where the issue with using` application_area_unit` as a dependency becomes because when it changes to a different unit the calculations are re-run with the new unit value which changes the UI string. 

So what I did is turn the dependency into a boolean, it only runs when `application_area_unit` changes from falsy to truthy (or vice versa). So it doesn’t run when application unit changes, lets say from ‘ac’ to ‘ft-2’ because both are truthy. 

Here's the code I used: 

```
  const previewStrings = useMemo(() => {
    let previewStringUnit;
    let previewStringValue;

    if (application_area_unit) {
      previewStringUnit =
        total_area_unit && location_area[system].units.includes(total_area_unit)
          ? getUnitOptionMap()[total_area_unit]
          : application_area_unit;

      previewStringValue =
        total_area &&
        roundToTwoDecimal(
          convertFn(
            location_area[system],
            total_area,
            location_area.databaseUnit,
            previewStringUnit.value,
          ),
        );
    }

    return { previewStringUnit, previewStringValue };
  }, [!!application_area_unit]);
```

Reply:
Mmmm I love the boolean value in dependency array! That is exactly what I wanted 😍

We won't be making any further changes in this PR to avoid merge conflicts with #3294 but I would like to adopt it there.

Reply:
Also just to mention the useEffect solution works fine so its not necessary to change it. I was just pointing out what I know and from reading the react docs, the exact same thing is mentioned in this page: https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state

Basically whenever something can be computed during a components render, then it probably shouldn't be in state. This is more like best practices and not a hard rule though.